(#581) Add recommendations for sxs packages (#587)
chocolatey/docs: input/en-us/create/create-packages.md

## Avoid folders named “content”

Do not use a folder named “content” in your package. NuGet attaches a special meaning to this folder and will not allow you to have dependencies on packages that have content folders without also having a content folder. It's turtles all the way down until we or NuGet removes this limitation.

## Naming your package

The **title** of your package (`<title>` tag in the nuspec) should be the same as the name of the application.
Follow the official spelling, use upper and lower case and don’t forget the spaces.
Examples of correct package titles are: _Google Chrome_, _CCleaner_, _PuTTY_ and _FileZilla_.
The title will appear on the left side in the package list of the Chocolatey gallery, followed by the version.

There are some guidelines in terms of the package **ID** (`<id>` tag in the nuspec):

* **Use only lowercase letters**, even if you used uppercase letters in the package title.
  (This is considered a guideline because it is correctable in other ways).

>   Once a package is submitted (even prior moderation), the Gallery will always show the id with the casing of the first package version.

Reviewer: ```suggestion
  Once a package is submitted (even prior moderation), the Chocolatey Community Repository will always show the id with the casing of the first package version.
```

* If the name of the application contains multiple words separated by spaces, such as _MusicBrainz Picard_ or _Adobe Reader_, replace the spaces with the hyphen-minus character “-” (U+002D), or just omit them.
  **Don’t use dots**.
  Dots should be used only if the original application name contains dots (for example, _Paint.NET_).
  Hence the correct id’s of the previously mentioned applications can be `musicbrainz-picard` or `adobereader`.
  We recommended you use the hyphen method when there are long package names to ensure they remain readable.
* For sub-packages, use the hyphen-minus character “-” (U+002D) as the separator, not a dot.
  Sub-packages are intended for separate packages that include extensions, modules or additional features/files for other applications.
  Therefore `keepass-langfiles` is a proper package id, because it adds the language files for the main application which in this case is _KeePass_.
  Another example is `libreoffice-help` for the help pack for _LibreOffice_, the open source office suite.

> * If you want to package an open source application, check the repositories of popular Linux distributions, such as [Debian](http://www.debian.org/distrib/packages#search_packages), [Ubuntu](http://packages.ubuntu.com/), or [Arch Linux](https://www.archlinux.org/packages/).

Reviewer 2: Should we maybe mention that they can search for package ideas on services like https://repology.org/? Instead of calling out specific distros?

Author: ```suggestion
* If you want to package an open source application, first search for any pre-existing packages on services like [Repology](https://repology.org/).
```

  If there are already published packages for the application on another repository, **use the same package id**.
  This will make it easier for users which work with multiple platforms, so they don't have to remember and use different package names.

These guidelines are already commonly applied on packages for all major Linux distributions, because they lead to a more consistent user experience for software repositories, result in easier to remember package IDs, and reduce unnecessary considerations on naming packages for package creators.

> Note that some packages in the Chocolatey Community Gallery don’t follow these guidelines; many of these were created before these guidelines were introduced.

Reviewer: ```suggestion
Note that some packages in the Chocolatey Community Repository don’t follow these guidelines; many of these were created before these guidelines were introduced.
```

Author: That weird quote is now bothering me so amending the suggestion to 'fix' that too lol

```suggestion
Note that some packages in the Chocolatey Community Repository don't follow these guidelines; many of these were created before these guidelines were introduced.
```

New packages should follow the guidelines to improve the overall experience for users going forward.

If you are going to offer a package that has both an installer and an archive (zip or executable only) version of the application, create three packages - see [Portable vs Installable](xref:faqs#what-distinction-does-chocolatey-make-between-an-installable-and-a-portable-application) and [Install, Portable, and Meta/Virtual Packages](xref:faqs#what-is-the-difference-between-packages-no-suffix-as-compared-to.install.portable).

### Naming packages to allow for side-by-side installation

For some software, it is expected that users may want multiple different versions installed simultaneously.
This has commonly been seen with software like Python which offer some kind of runtime dependency that other scripts or applications depend on.

> In the past, Chocolatey supported a "side-by-side" installation method which could be used for this purpose to install multiple distinct versions side-by-side on the same machine.

Reviewer: ```suggestion
In the past, Chocolatey CLI supported a "side-by-side" installation method which could be used for this purpose to install multiple distinct versions side-by-side, on the same machine.
```

> This has since been deprecated and will be removed in a future version due to many issues with the design of the feature.

Reviewer: ```suggestion
This has since been deprecated and will be removed in version 2.0.0 due to many issues with the use and implementation of the feature.
```

> Other features may take its place in the future, but for now we instead recommend you try one of the below solutions, depending on your use case.

Reviewer: ```suggestion
```

#### If the software you're creating a package for expects **side-by-side installations of different versions**

1. Create _separate package IDs_ for each major version (or major + minor version, depending on the expectations of users) which include the version number in the ID.

>   For example, if the package ID would be `python`, the package for major version 2 would be `python2`, or a package for specifically the 2.7 branch might be `python2.7`.

Reviewer: ```suggestion
  For example, if the package ID would be `python`, the package for major version 2 would be `python2`, or a package for specifically the 2.7 branch might be `python27` or `python2-7`.
```

Reviewer: This is to adhere to the guidelines of using dots in package ID's.

  These individual versioned packages can be updated as normal, as long as that major and/or minor version is still receiving further updates.
2. Create a [**metapackage**](xref:getting-started#terminology) for the overall product with a package version for each of the versioned packages.

>   In other words, create an empty package that contains only a `nuspec` file under the package ID without a version number (in the above example, the metapackage would be called `python`), and have it depend on the versioned package.
>   Following the above example for Python, we'd create a package with the ID `python` at version `2` that takes a dependency on the `python2` package.

Reviewer: ```suggestion
  In other words, create an empty package, named as just the package ID without a version number, that contains only a `nuspec` file and add a dependency on the versioned packagehave it depend on the versioned package. 
  Following the above example for Python, we'd create a package with the ID `python` at version `2.0.0` that takes a dependency on the `python2` package.
```

Author: @pauby I think there's some stray text in the suggestion there that wasn't removed. This sentence is also getting very awkwardly phrased now I come back to it... I think this captures what we were going for, maybe?

```suggestion
  In other words, create another package (this time without a version number in the ID), which contains only a `nuspec` file and add a dependency on the versioned package.
  Following the above example for Python, we'd create a package with the ID `python` at version `2.0.0` that takes a dependency on the `python2` package.
```

3. New versions of the metapackage should be published for "sub-versions" of the dependency package as well.

>   For example, if Python version `3` is released, we'd create a new version of the `python` package with its version set to `3.0`, which depends on a new `python3` package.

Reviewer: ```suggestion
  For example, if Python version `3` is released, we'd create a new version of the `python` package with its version set to `3.0.0`, which depends on a new `python3` package.
```

#### If the software you're creating a package for expects **side-by-side installations of the same version**

A possible example might be where multiple different instances of the software are being installed into different install directories.

Chocolatey does not currently support this functionality.

> This functionality can be mimicked by making use of `--force` to forcibly reinstall an already-installed package and providing an argument or package parameter that the package can use to direct the installation to install into a different directory.

Reviewer 2: Since we have mentioned in the past that `--force` isn't a reinstall, but rather replacing the installed package, should we be calling this a reinstall in the documentation? (See https://github.com/chocolatey/choco/issues/2724)

Reviewer: Agreed. We should change that.

Author: ```suggestion
This functionality can be mimicked by making use of `--force` to forcibly overwrite an already-installed package and providing an argument or package parameter that the package can use to direct the installation to install into a different directory.
```

However, Chocolatey will only be aware of the most-recently installed copy of the software, and will only be able to provide updates to that single installation of the software in most cases.
As a result, this particular configuration is not supported.

A potential workaround for users maintaining their own package repositories is to maintain separate instances of the same package under different package IDs, where each package ships the same software, but is scripted to install to a different directory.

> Recommended naming conventions here would be similar to sub-packages, where the package is named after the primary package ID followed by a dash and a discriminator, for example `packageID-secondary`.

Reviewer: A solution to this could be to create a new package with a different ID. Taking python3 as an example you could have a `python3-localtoolsfolder` package and a `python3-programfilesfolder` package. They would effectively be duplicates of each other which just slightly different parameters passed to the install function but it is a solution, not a good one, but it is one?

Author: This is more or less what I said here I think, but your version is a bit more clear. I do consider it more a workaround than a solution, though, because it effectively decouples  the package from any upstream updates which you then have to handle manually / figure out your own automation for.

Let me see if I can incorporate some of what you have here though, it's phrased a bit better than what I had for this. Thanks!
